validator/src/commands/exit/mod.rs

```diff
@@ -102,16 +104,60 @@ pub fn execute(matches: &ArgMatches, ledger_path: &Path) -> Result<()> {
     }
 
     let admin_client = admin_rpc_service::connect(ledger_path);
-    admin_rpc_service::runtime().block_on(async move { admin_client.await?.exit().await })?;
+    let validator_pid =
+        admin_rpc_service::runtime().block_on(async move { admin_client.await?.exit().await })?;
     println!("Exit request sent");
 
     if exit_args.monitor {
         monitor::execute(matches, ledger_path)?;
+    } else {
+        poll_until_pid_terminates(validator_pid);
     }
 
     Ok(())
 }
 
+#[cfg(target_os = "linux")]
+// This function will be called while the process with id `pid` is terminating.
+// Inspecting  a process that is on the verge of termination is inherently racy
+// so the error handling and flow control reflect that
+fn poll_until_pid_terminates(pid: u32) {
+    // TODO: i32::try_from()
+    let pid = pid as i32;
+    println!("Monitoring agave-validator process {pid}");
+    loop {
+        let process = match Process::new(pid) {
```

Comment:
i think there's something under `wait(2)` that is more direct here. one of `wait{,p}id` with `WNOHANG` or something

Reply:
The interface does look more direct than what I'm doing, but I'm not positive this will work for us. From `man wait`:
```
All of these system calls are used to wait for state
changes in a child of the calling process,
and obtain information about the child whose state has changed.
```
Our issue is that we have to do the waiting on the "client side" as exit will tear down the `AdminRpc` handler that would send a response back to the client

Reply:
assuming this is what's under the hood of the core-utils `wait` command. iirc i've used it to wait nonchildren

Reply:
derp n/m 
![Screenshot from 2025-05-13 22-06-22](https://github.com/user-attachments/assets/80ef3d31-a004-4658-ac96-ea188767fb1f)


looks like you _can_ do `kill(pid, 0)` or pidfn_open(2)

Reply:
> looks like you can do kill(pid, 0) or pidfn_open(2)

Just to make sure I follow, we'd still need to call each of these in a loop and then bail when we get the "process not found" error (`ESRCH` for kill for example) right ? Are your rec's aimed at avoiding the introduction of an extra dep in favor of just using a libc function directly ?

Reply:
Getting the number of alive threads serves as a "progress bar", and the snippet I posted above is useful to debug which threads aren't bailing. It would be a little more difficult to get the number of non-terminated children threads without

Reply:
> > looks like you can do kill(pid, 0) or pidfn_open(2)

> Just to make sure I follow, we'd still need to call each of these in a loop and then bail when we get the "process not found" error (`ESRCH` for kill for example) right ?

yep

> Are your rec's aimed at avoiding the introduction of an extra dep in favor of just using a libc function directly ?

yeah no new dep and no fs access

Reply:
> Getting the number of alive threads serves as a "progress bar", and the snippet I posted above is useful to debug which threads aren't bailing. It would be a little more work to get the number of non-terminated children threads with one of the methods you described

how confident are you that the thread count is monotonically decreasing on exit? :trollface: